Adds GC feature to controller

sonnysideup opened the pull request:
* Adds CLI flags related to configurable GC settings (i.e. maxRetention, loop interval).
* Allows for disabling feature
* Creates a separate goroutine alongside standard controller reconciliation flow to handle the deletion of `ContainerImageBuild` resources
* Adds decent info/debug logging to GC operation

sonnysideup committed:
adds RunGC func

and a hook to start gc loop when controller is started

sonnysideup commented:
@steved please let me know what you think about this design at a high-level (defaults, design, etc...). I know there's a `FakeClient` we can leverage to write unit tests against this so I'm thinking I'd like to use that to exercise and map out the expected behaviour.

Also, I don't believe I need to pass a _stop signal_ to the goroutine I create but let me know if you disagree. It should be terminated just fine when the main routine exits and there's no real in-flight state that could cause a race condition that comes readily to mind.

ddl-audi reviewed (approved):
Looks fine to me. Endless loop is as good a design as any IMO

steved commented on controllers/start.go:
```diff
@@ -46,18 +47,31 @@ func StartManager(cfg ControllerConfig) {
 
+	if cfg.GCInterval != 0 {
+		go func() {
+			for {
```
T(some time last week)IL about https://golang.org/pkg/time/#Ticker

sonnysideup replied:
OoOoOoO...you fancy. Let me check it out.

sonnysideup replied:
https://github.com/dominodatalab/forge/pull/49/commits/71d838a93d5843b5aeb4edd062c3fe63bfb8503d

steved commented on controllers/containerimagebuild_controller.go:
```diff
@@ -99,3 +103,51 @@ func (r *ContainerImageBuildReconciler) Reconcile(req ctrl.Request) (ctrl.Result
+	})
+
+	for _, build := range builds[:len(builds)-retentionCount] {
+		if err := r.Delete(ctx, &build); err != nil {
```
Background or foreground? I had this debate with myself inside domino, but I'm curious your thoughts. I think the default is to block and there's no real performance concerns since this loop runs infrequently.

sonnysideup replied:
Foreground for now -> https://github.com/dominodatalab/forge/pull/49/commits/c3933061c019b0b83c3b42bc7e3c1767043ff1d6

steved commented on cmd/root.go:
```diff
@@ -171,6 +177,8 @@ func init() {
 	rootCmd.Flags().StringVar(&buildJobSecurityContextConstraints, "build-job-security-context-constraints", "", "Run builds jobs using a specified SCC")
 	rootCmd.Flags().BoolVar(&buildJobGrantFullPrivilege, "build-job-full-privilege", false, "Run builds jobs using a privileged root user")
 	rootCmd.Flags().StringVar(&buildAdvancedConfigFilename, "build-job-advanced-config", "", "Add volumes, volume mounts and environment variables to your build jobs using a JSON file")
+	rootCmd.Flags().DurationVar(&gcInterval, "gc-interval", 30*time.Minute, "Run ContainerImageBuild cleanup operation according to this interval. Set to 0 to disable")
+	rootCmd.Flags().IntVar(&gcMaxKeepCount, "gc-max-keep", 5, "Delete all ContainerImageBuild resources in a 'finished' state that exceed this count")
```
I actually don't think we've spoken about this, but I removed the TTL on the job in https://github.com/dominodatalab/forge/commit/23541dd3563e8d64d03fa9403aced3ba20f2847a#diff-dd01fb2f9549977a5fa7fb3c85839ec7L277. Mostly because an hour is quite low, but I do wonder if we should set a high one (like 1 day or something) or if we need something like this to clean those up.

sonnysideup replied:
The question that comes immediately to mind is, "What does a terminated Pod cost?". I don't think it's much beyond a minor footprint of metadata storage in etcd; I doubt K8s keeps the Pod's former resource allocation locked up but I don't know for sure.

Obviously, when we nuke a `ContainerImageBuild` the owned `Job` and, by extension, the `Pod` the job owns will be deleted. We probably don't need this field and I added it just because.

It might also be worth removing this setting as well. The TTLController feature is in _alpha_, requires a gate setting to enable it right now, and, we should probably limit our delete process to using the owner `ContainerImageBuild` resource for consistency's sake, no?

https://github.com/dominodatalab/forge/blob/a87afe1dcf1a4f903cd5d1286f1552c8f0ad5ea0/controllers/containerimagebuild_resources.go#L239

steved replied:
Oh interesting. Yeah it seems like we should operate on CIBs not jobs.

The secondary point is mostly around whether we should handle a potential CIB that is still running after some time that we deem excessive. i.e. if we have a build that just executes `sleep infinity` is that reasonable? Probably, since it's on the customer to figure out why they would do that. Or, we can set an upper limit and forcefully clean them up here.

As I was writing this however, I think I talked myself out of any sort of "active deadline" default at all for running CIBs.

sonnysideup replied:
I agree, clients should handle user garbage like this. I think our responsibility is to prevent the accumulation of _finished_ cruft, not assume long-running builds are invalid and nuke them.

steved reviewed (commented):
Design & defaults-wise LGTM. Would love to see a fake.Client (or suite) test :+1:

sonnysideup committed:
updates vendored deps

it sucks but we have to run `go mod vendor` every time we update the
module file

sonnysideup committed:
adds unit tests for RunGC func

sonnysideup committed:
changes gc routine to delete in "foreground"

sonnysideup committed:
uses ticker for triggering GC instead of sleep

channels are your friend

sonnysideup committed:
removes TTLSecondsAfterFinished field from buildjob

we're going to limit the delete pathways to one source: the parent CIB

sonnysideup committed:
changes project ownership

cgk owns this now

sonnysideup commented:
@steved This is ready to ship as-is I think but I wanted to mention a couple of things.

1. The tests using the fake client are in-place and exercise everything except the error modes. In order to trigger an error and build test expectations, I'd have to type-embed `client.Client` into a custom type and use the fake client underneath the hood. Since we don't return an error nor do we panic explicitly, I'm not sure if that's worth it. The cost: these tests are not 100% exhaustive.  Thoughts?
2. Do you think we should _emit_ a failure event into K8s if an error occurs during GC? With the way this is written, the only way to tell if an error occurs is to check pod logs so we could surface this into the k8s events api as well; doesn't cost much I already have an event recorder embedded into the reconciler.

steved commented:
> The tests using the fake client are in-place and exercise everything except the error modes. In order to trigger an error and build test expectations, I'd have to type-embed client.Client into a custom type and use the fake client underneath the hood. Since we don't return an error nor do we panic explicitly, I'm not sure if that's worth it. The cost: these tests are not 100% exhaustive. Thoughts?

> Do you think we should emit a failure event into K8s if an error occurs during GC? With the way this is written, the only way to tell if an error occurs is to check pod logs so we could surface this into the k8s events api as well; doesn't cost much I already have an event recorder embedded into the reconciler.

I guess these two points are a little connected, in that if you add support for the k8s event you probably want to test it. I do like the idea of an event, though is it possible it'd spam new events if an object is undeletable? Did you mean an event on the CIB itself?

steved commented on controllers/start.go:
```diff
@@ -46,18 +47,37 @@ func StartManager(cfg ControllerConfig) {
+		}()
+
+		go func() {
+			select {
+			case <-ticker.C:
```
I think you can `for range ticker.C`?

sonnysideup committed:
uses for range for gc ticker

sonnysideup commented:
> I guess these two points are a little connected, in that if you add support for the k8s event you probably want to test it. I do like the idea of an event, though is it possible it'd spam new events if an object is undeletable? Did you mean an event on the CIB itself?

I don't think there's a way to have the CIB emit events, only the reconciler. It's true that, if we find ourselves in a GC loop that cannot list nor delete CIBs, then this would emit events continually but I think the _spamming_ would only occur if the interval is set to something low (not the default).

Consider the following:

* If `RunGC` returns an error, how do we handle that? I'm not sure there's anything we can do which is why I don't return one.
* Should we `panic` and kill the process altogether? Probably not since our main concern is process builds.
* Should we report the error? Yes, and that's currently done in the logs but how to do we monitor that in a clean way? Log processing is, at best, mildly gross. I was thinking that events, since that can be of type `EventTypeWarning`, open up a concrete way to monitor for issues (i.e. something can watch events and process warnings).

Your points re: testing are valid. Let me take one more look at this and see if I can easily get that tested. Does this make sense?

sonnysideup committed:
emits events during errors in GC routine

and updates tests to be exhaustive
